pallets/parachain-staking/src/lib.rs

```diff
@@ -0,0 +1,1322 @@
+
+pub use pallet::*;
+
+#[pallet]
+pub mod pallet {
+
+	use super::InflationInfo;
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{Currency, Get, Imbalance, LockIdentifier, LockableCurrency, ReservableCurrency, WithdrawReasons},
+	};
+	use frame_system::pallet_prelude::*;
+	// TODO: Use ORML one once they point to Substrate master
+	// use orml_utilities::OrderedSet;
+	use crate::set::OrderedSet;
+	use parity_scale_codec::{Decode, Encode};
+	use sp_runtime::{
+		traits::{AtLeast32BitUnsigned, Saturating, Zero},
+		Perbill, RuntimeDebug,
+	};
+	use sp_std::{cmp::Ordering, collections::btree_map::BTreeMap, prelude::*};
+
+	pub const REWARDS_ID: LockIdentifier = *b"kiltrwrd";
+
+	/// Pallet for parachain staking
+	#[pallet::pallet]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[derive(Default, Clone, Encode, Decode, RuntimeDebug)]
+	pub struct Bond<AccountId, Balance> {
+		pub owner: AccountId,
+		pub amount: Balance,
+	}
+
+	impl<A, B: Default> Bond<A, B> {
+		fn from_owner(owner: A) -> Self {
+			Bond {
+				owner,
+				amount: B::default(),
+			}
+		}
+	}
+
+	impl<AccountId: Ord, Balance: PartialEq> Eq for Bond<AccountId, Balance> {}
+
+	impl<AccountId: Ord, Balance: PartialEq> Ord for Bond<AccountId, Balance> {
+		fn cmp(&self, other: &Self) -> Ordering {
+			self.owner.cmp(&other.owner)
+		}
+	}
+
+	impl<AccountId: Ord, Balance: PartialEq> PartialOrd for Bond<AccountId, Balance> {
+		fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+			Some(self.cmp(other))
+		}
+	}
+
+	impl<AccountId: Ord, Balance: PartialEq> PartialEq for Bond<AccountId, Balance> {
+		fn eq(&self, other: &Self) -> bool {
+			self.owner == other.owner && self.amount == other.amount
+		}
+	}
+
+	#[derive(Copy, Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
+	/// The activity status of the collator
+	pub enum CollatorStatus {
+		/// Committed to be online and producing valid blocks (not equivocating)
+		Active,
+		/// Temporarily inactive and excused for inactivity
+		Idle,
+		/// Bonded until the inner round
+		Leaving(RoundIndex),
+	}
+
+	impl Default for CollatorStatus {
+		fn default() -> CollatorStatus {
+			CollatorStatus::Active
+		}
+	}
+
+	#[derive(Default, Encode, Decode, RuntimeDebug)]
+	/// Snapshot of collator state at the start of the round for which they are
+	/// selected
+	pub struct CollatorSnapshot<AccountId, Balance> {
+		pub bond: Balance,
+		pub delegators: Vec<Bond<AccountId, Balance>>,
+		pub total: Balance,
+	}
+
+	impl<AccountId: Ord, Balance: PartialEq> PartialEq for CollatorSnapshot<AccountId, Balance> {
+		fn eq(&self, other: &Self) -> bool {
+			self.bond == other.bond && self.total == other.total && self.delegators.eq(&other.delegators)
+		}
+	}
+
+	impl<AccountId: Ord, Balance: PartialEq> Eq for CollatorSnapshot<AccountId, Balance> {}
+
+	#[derive(Encode, Decode, RuntimeDebug)]
+	/// Global collator state with commission fee, bonded stake, and delegations
+	pub struct Collator<AccountId, Balance> {
+		pub id: AccountId,
```

Comment:
with this much type definition code, might be cool to refactor into a `types.rs`